test/core/test/hooks-stack.test.ts

```diff
@@ -0,0 +1,47 @@
+import * as vitest from 'vitest'
+import { describe, expect, test, vi } from 'vitest'
+
+vi.setConfig({
+  sequence: {
+    hooks: 'stack',
+  },
+})
+
+const hookOrder: number[] = []
+function callHook(hook: 'beforeAll' | 'beforeEach' | 'afterAll' | 'afterEach', order: number) {
+  vitest[hook](() => {
+    hookOrder.push(order)
+  })
+}
+
+describe('hooks are called sequentially', () => {
```

Comment:
This describes correct behavior and it works as far as I can tell. A possible change could be to write it like this:
```ts
describe('hooks are called sequentially', () => {
  callHook('beforeAll', 1)
  callHook('beforeAll', 2)
  callHook('beforeAll', 3)

  callHook('afterAll', 4)
  // will wait for it
  vitest.afterAll(async () => {
    await Promise.resolve()
    hookOrder.push(5)
  })
  callHook('afterAll', 6)

  callHook('beforeEach', 7)
  callHook('afterEach', 8)

  callHook('beforeEach', 9)
  callHook('afterEach', 10)

  callHook('beforeEach', 11)
  callHook('afterEach', 12)

  test('before hooks pushed in order', () => {
    expect(hookOrder).toEqual([1, 2, 3, 7, 9, 11])
  })
})

describe('previous suite run all hooks', () => {
  test('after all hooks run in reverse order', () => {
    expect(hookOrder).toEqual([1, 2, 3, 7, 9, 11, 12, 10, 8, 6, 5, 4])
  })
})
```

This emphasizes the feature that I am looking for: the ability to group `before*` and `after*` hooks so they can be extracted into functions that can be reused. But definitely not a show-stopper.